ffs/scheduler/scheduler_storage.go

```diff
@@ -203,34 +204,47 @@ func (s *Scheduler) StorageConfig(jid ffs.JobID) (ffs.StorageConfig, error) {
 // should be considered failed. If error is nil, it still can return []ffs.DealError
 // since some deals failing isn't necessarily a fatal Job config execution.
 func (s *Scheduler) executeStorage(ctx context.Context, a astore.StorageAction, job ffs.StorageJob, dealUpdates chan deals.StorageDealInfo) (ffs.StorageInfo, []ffs.DealError, error) {
-	ci, err := s.getRefreshedInfo(ctx, a.Cid)
+	ci, err := s.getRefreshedInfo(ctx, a.APIID, a.Cid)
 	if err != nil {
 		return ffs.StorageInfo{}, nil, fmt.Errorf("getting current cid info from store: %s", err)
 	}
 
 	if a.ReplacedCid.Defined() {
-		if err := s.Untrack(a.ReplacedCid); err != nil && err != astore.ErrNotFound {
+		if err := s.Untrack(a.APIID, a.ReplacedCid); err != nil && err != astore.ErrNotFound {
 			return ffs.StorageInfo{}, nil, fmt.Errorf("untracking replaced cid: %s", err)
 		}
 	}
 
-	s.l.Log(ctx, "Ensuring Hot-Storage satisfies the configuration...")
-	hot, err := s.executeHotStorage(ctx, ci, a.Cfg.Hot, a.Cfg.Cold.Filecoin.Addr, a.ReplacedCid)
-	if err != nil {
-		s.l.Log(ctx, "Hot-Storage excution failed.")
-		return ffs.StorageInfo{}, nil, fmt.Errorf("executing hot-storage config: %s", err)
+	var hot ffs.HotInfo
+	if a.Cfg.Hot.Enabled {
+		s.l.Log(ctx, "Executing Hot-Storage configuration...")
+		hot, err = s.executeEnabledHotStorage(ctx, a.APIID, ci, a.Cfg.Hot, a.Cfg.Cold.Filecoin.Addr, a.ReplacedCid)
+		if err != nil {
+			s.l.Log(ctx, "Enabled Hot-Storage excution failed.")
+			return ffs.StorageInfo{}, nil, fmt.Errorf("executing enabled hot-storage: %s", err)
+		}
+		s.l.Log(ctx, "Hot-Storage configuration ran successfully.")
 	}
-	s.l.Log(ctx, "Hot-Storage execution ran successfully.")
 
-	s.l.Log(ctx, "Ensuring Cold-Storage satisfies the configuration...")
+	s.l.Log(ctx, "Executing Cold-Storage configuration...")
 	cold, errors, err := s.executeColdStorage(ctx, ci, a.Cfg.Cold, dealUpdates)
 	if err != nil {
 		s.l.Log(ctx, "Cold-Storage execution failed.")
 		return ffs.StorageInfo{}, errors, fmt.Errorf("executing cold-storage config: %s", err)
 	}
-	s.l.Log(ctx, "Cold-Storage execution ran successfully.")
+	s.l.Log(ctx, "Cold-Storage configuration ran successfully.")
+
+	if !a.Cfg.Hot.Enabled {
+		s.l.Log(ctx, "Executing Hot-Storage configuration...")
+		if err := s.executeDisabledHotStorage(ctx, a.APIID, a.Cid); err != nil {
+			s.l.Log(ctx, "Disabled Hot-Storage execution failed.")
+			return ffs.StorageInfo{}, nil, fmt.Errorf("executing disabled hot-storage: %s", err)
+		}
+		s.l.Log(ctx, "Hot-Storage configuration ran successfully.")
+	}
```

Comment:
If HotStorage wasn't enabled, now run hot-storage execution which will unpin the Stage-pin potentially set by a staging flow.

To be clear:
- If HotStorage is enabled, this code first executes HotStorage logic since it will promote the Staged-pin to a full pin... so it's still safe to run ColdStorage afterwards (since the data is still pinned).
- If HotStorage is disabled, we avoid running HotStorage logic first since that would unpin the data and potentially make ColdStorage logic be unsafe to run correctly (if IPFS GC runs in the middle). So we run it afterwards so it can unpin the data safely since cold storage things finished already.